runtime/decimal.c

```diff
@@ -3,49 +3,57 @@
 
 typedef GC decQuad *DecimalPtr;
 
-typedef struct {
-    TaggedPtr ptr;
-    PanicCode panicCode;
-} TaggedPtrPanicCode;
-
 // Status flags that we should treat as failure
 // XXX need to investigate DEC_Underflow
 #define DECIMAL_STATUS_FAIL (DEC_Errors|DEC_Subnormal)
 
-static inline PanicCode statusPanicCode(uint32_t status) {
-    if (status & DECIMAL_STATUS_FAIL) {
-        // XXX decide what panic codes to use 
-        return (status & DEC_Division_by_zero) ? PANIC_DIVIDE_BY_ZERO : PANIC_ARITHMETIC_OVERFLOW;
-    }
-    return 0;
-}
-
 static inline TaggedPtr createDecimal(decQuad *dq) {
     DecimalPtr dp = _bal_alloc(sizeof(decQuad));
     *dp = *dq;
-    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL);
+    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL << TAG_SHIFT);
 }
 
-static inline decQuad *taggedToDecQuad(TaggedPtr tp) {
+static inline TaggedPtrPanicCode finish(decQuad *dq, decContext *cx) {
```

Comment:
Since we're now calling decQuadClass, let's not inline this.

Reply:
Fixed